### DISCUSSION TITLE
Multisig smart wallet unable to call function of another contract.

### DISCUSSION THREAD

**Discussion by Nadhiem999** [zkSync-Community-Hub/zksync-developers#458] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

i am trying to call a function of another contract after deploying aa multisig account. 

### What is the specific issue or error you're encountering?

i am getting this error when calling the function  message: "invalid sender. can't start a transaction from a non-account",

### Can you share the error messages or logs you're receiving, if any?

invocation: null,
  revert: null,
  shortMessage: 'missing revert data',
  info: {
    error: {
      code: 3,
      message: "invalid sender. can't start a transaction from a non-account",
      data: '0x'
    },
    payload: {
      method: 'eth_estimateGas',
      params: [Array],
      id: 2,
      jsonrpc: '2.0'
    }
  }
}

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

async function addLiquidityThroughMultisig() {

  
    
 //   const result = await ammContract.addLiquidity(tokenAAddress, tokenBAddress, amountA, amountB);
 
    let aaTx = await ammContract.addLiquidity.populateTransaction(tokenAAddress, tokenBAddress, amountA, amountB);

    const gasLimit = await provider.estimateGas(aaTx);
    const gasPrice = await provider.getGasPrice();
    const chainId = (await provider.getNetwork()).chainId;
    const nonce = await provider.getTransactionCount(multisigAddress);
    console.log(multisigAddress)
     
    aaTx = {
        ...aaTx,
        from: multisigAddress,
        gasLimit: gasLimit,
        gasPrice: gasPrice,
        chainId: chainId,
        nonce: nonce,
        type: 113,
        customData: {
            gasPerPubdata: zk.utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
        } as zk.types.Eip712Meta,
        value: ethers.BigNumber.from(0),
    }

    const signedTxHash = zk.EIP712Signer.getSignedDigest(aaTx)

    const signature = ethers.utils.concat([
        // Note, that `signMessage` wouldn't work here, since we don't want
        // the signed hash to be prefixed with `\x19Ethereum Signed Message:\n`
        ethers.utils.joinSignature(owner1._signingKey().signDigest(signedTxHash)),
        ethers.utils.joinSignature(owner2._signingKey().signDigest(signedTxHash)),
    ])

    // Check the Signature is Valid
    const isValidSignature = await aa.isValidSignature(signedTxHash, signature)
    if (isValidSignature == EIP1271_MAGIC_BYTES) {
        console.info(chalk.green(`Signaure is Valid in MultiSigAccount!`))
    } else {
        throw new Error(`Signature is not valid in MultiSigAccount!`)
    }

    aaTx.customData = {
        ...aaTx.customData,
        customSignature: signature,
    }

    console.log(
        `The multisig's nonce before the first tx is ${await provider.getTransactionCount(
            multisigAddress,
        )}`,
    )
    const sentTx = await provider.sendTransaction(zk.utils.serialize(aaTx))
    await sentTx.wait()

    // Checking that the nonce for the account has increased.
    console.log(
        `The multisig's nonce after the first tx is ${await provider.getTransactionCount(
            multisigAddress,
        )}`,
    )

    
    // }

        
}
addLiquidityThroughMultisig().catch(console.error);

### Have you tried to isolate the problem, and if so, what were the results?

still the same issue

### What steps have you already taken to try to resolve the issue?

i tried everything possible i could find in documentation

**Comment by uF4No** [comment]
That error is returned when the contract sending the transaction (in this case the smart wallet) was not deployed using `createAccount` or `create2Account`. The protocol does not know that this contract is actually an account and returns this error.

As an example, you can see the [AccountFactory from our multisig tutorial](https://docs.zksync.io/build/tutorials/smart-contract-development/account-abstraction/custom-aa-tutorial.html#the-factory) that uses `create2Account` (see line 25).

**Reply by Nadhiem999** [reply]
Thank you for the reply . Here are the step : 
1 . Deployed AA factory 
   it has the function which uses createAccount 
   function deployAccount(
        bytes32 salt,
        address owner1,
        address owner2
    ) external returns (address accountAddress) {
        (bool success, bytes memory returnData) = SystemContractsCaller
            .systemCallWithReturndata(
                uint32(gasleft()),
                address(DEPLOYER_SYSTEM_CONTRACT),
                uint128(0),
                abi.encodeCall(
                    DEPLOYER_SYSTEM_CONTRACT.create2Account,
                    (salt, aaBytecodeHash, abi.encode(owner1, owner2), IContractDeployer.AccountAbstractionVersion.Version1)
                )
            );
         2. Used AAfactory deployAccount  function to make a wallet and fund it . i was able to fund it.
         3. when i try to call a function using this smart  wallet  i get the error  message: "invalid sender. can't start a transaction from a non-account"

**Reply by Nadhiem999** [reply]
this is how i am trying to call the function 
 let aaTx = await ammContract.populateTransaction.addLiquidity.(tokenAAddress, tokenBAddress, amountA, amountB);

    const gasLimit = await provider.estimateGas(aaTx);
    const gasPrice = await provider.getGasPrice();
    const chainId = (await provider.getNetwork()).chainId;
    const nonce = await provider.getTransactionCount(multisigAddress);
    console.log(multisigAddress)
     
    aaTx = {
        ...aaTx,
        from: multisigAddress,
        gasLimit: gasLimit,
        gasPrice: gasPrice,
        chainId: chainId,
        nonce: nonce,
        type: 113,
        customData: {
            gasPerPubdata: zk.utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
        } as zk.types.Eip712Meta,
        value: ethers.BigNumber.from(0),
    }

    const signedTxHash = zk.EIP712Signer.getSignedDigest(aaTx)

    const signature = ethers.utils.concat([
        // Note, that `signMessage` wouldn't work here, since we don't want
        // the signed hash to be prefixed with `\x19Ethereum Signed Message:\n`
        ethers.utils.joinSignature(owner1._signingKey().signDigest(signedTxHash)),
        ethers.utils.joinSignature(owner2._signingKey().signDigest(signedTxHash)),
    ])

**Reply by Nadhiem999** [reply]
i have even tried a couple of things : 
1. tried to call a basic contract of greeter that is in zksync tutorials , which would still give the same error .
here is the script  i am using for interaction 
`let aaTx = await greeterContract.setGreeting.populateTransaction("Hello from multisig!");
 console.log(aaTx)
  const gasLimit = await provider.estimateGas({ ...aaTx, from: multisigAddress });
 // const gasLimit= await provider.estimateGas(aaTx);
  console.log(gasLimit)
  const gasPrice = await provider.getGasPrice();

  aaTx = {
    ...aaTx,
    from: multisigAddress,
    gasLimit: gasLimit,
    gasPrice: gasPrice,
    chainId: (await provider.getNetwork()).chainId,
    nonce: await provider.getTransactionCount(multisigAddress),
    type: 113,
    customData: {
      gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
    } as types.Eip712Meta,
    value: 0n,
  };

  const signedTxHash = EIP712Signer.getSignedDigest(aaTx);
  console.log(signedTxHash)

  // Sign the transaction with both owners
  const signature = ethers.concat([ethers.Signature.from(owner1.signingKey.sign(signedTxHash)).serialized, ethers.Signature.from(owner2.signingKey.sign(signedTxHash)).serialized]);

  aaTx.customData = {
    ...aaTx.customData,
    customSignature: signature,
  };


  const sentTx = await provider.broadcastTransaction(types.Transaction.from(aaTx).serialized);
  console.log(`Transaction sent from multisig with hash ${sentTx.hash}`);

 

  await sentTx.wait();`
  Meanwhile also , what are the scenarios which could lead to this error 'message: 'failed to validate the transaction. reason: ECDSA: invalid signature' out of curiosity.

**Reply by dutterbutter** [reply]
Hey @Nadhiem999 can you provide a link to the repo so its easier to reproduce / debug?

**Reply by Nadhiem999** [reply]
here is the link : https://github.com/Nadhiem999/multisig/blob/main/deploy/deploy-multisig.ts.

error : shortMessage: 'missing revert data',
  info: {
    error: {
      code: 3,
      message: 'failed to validate the transaction. reason: ECDSA: invalid signature',
      data: '0x'
    },
    payload: {
      method: 'eth_estimateGas',
      params: [Array],
      id: 49,
      jsonrpc: '2.0'
    }
  }
} 

and while trying to interact with greeter 
invocation: null,
revert: null,
shortMessage: 'missing revert data',
info: {
error: {
code: 3,
message: "invalid sender. can't start a transaction from a non-account",
data: '0x'
},
payload: {
method: 'eth_estimateGas',
params: [Array],
id: 2,
jsonrpc: '2.0'
}
}
}

**Reply by dutterbutter** [reply, marked as answer]
I reviewed the link provided here are my suggested fixes as I was able to execute the [tx](https://sepolia.explorer.zksync.io/tx/0x5d3be6863f93bc66669ac863131854856b66b1325023c4bf20cd0040a8f61592) from your repo. 

In package.json update your dependencies as such:

- Remove the following:
```
"@matterlabs/hardhat-zksync-deploy": "^1.1.2",
"@matterlabs/hardhat-zksync-node": "^1.0.1",
"@matterlabs/hardhat-zksync-solc": "^1.0.6",
"@matterlabs/hardhat-zksync-verify": "^1.2.2",
"zksync-web3": "^0.17.1"
```

- Add the following:
```
 "@matterlabs/hardhat-zksync": "^1.0.0",
```

- Update the following:
```
"zksync-ethers": "^6.7.0"
```

In your `deploy-multisig.ts` file make use of the `SmartAccount` class and leverage the `MultisigECDSASmartAccount`. This will simplify multisig usage significantly. For example:

```
let aaTx = await ammContract.getPoolId.populateTransaction(tokenAAddress, tokenBAddress);
const multisigAccount = MultisigECDSASmartAccount.create(
  multisigAddress,
  [owner1.signingKey, owner2.signingKey],
  provider,
);
const txx = await multisigAccount.sendTransaction({ ...aaTx });
await txx.wait();
console.log("Transaction sent from multisig smart account", txx.hash);
```

Output:
```
npm run hardhat deploy-zksync --script deploy-multisig.ts
Multisig account deployed on address 0x3AF4402E0DdFB55530E93fd004442544F4d7224D
Sending funds to multisig account...
Transaction sent from multisig 0x5d3be6863f93bc66669ac863131854856b66b1325023c4bf20cd0040a8f61592
```
